Added suite connector integration test

nzhechev-bosch opened the pull request:
[#44] Add suite connector integration test

Signed-off-by: Nikola Zhechev <Nikola.Zhechev@bosch.io>

nzhechev-bosch committed:
[#44] Added suite connector integration test

Signed-off-by: Nikola Zhechev <Nikola.Zhechev@bosch.io>

nzhechev-bosch committed:
[#44] Fixed golint errors

Signed-off-by: Nikola Zhechev <Nikola.Zhechev@bosch.io>

nzhechev-bosch committed:
[#44] Added delta to readySince/Until validity checks

Signed-off-by: Nikola Zhechev <Nikola.Zhechev@bosch.io>

nzhechev-bosch committed:
[#44] Fixed time delta config property name

Signed-off-by: Nikola Zhechev <Nikola.Zhechev@bosch.io>

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,430 @@
+// Copyright (c) 2022 Contributors to the Eclipse Foundation
```
Could you please apply the new project copyright header?

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,430 @@
+	MqttQuiesceMs            int    `def:"500"`
+	MqttAcknowledgeTimeoutMs int    `def:"3000"`
+
+	DittoHost string
```
Just an idea to keep the names independent of the concrete Eclipse project names. For example, Hono/Ditto can be integrated in another deployment i.e. the connection will be to the wrapper project.

As a reference, in the documentation, we decided just to say that suite connector connects MQTT endpoint and the configuration property is just `address`.

nzhechev-bosch replied:
This is aligned with https://github.com/eclipse-kanto/kanto/pull/127, which reuses the same environmental variable.
Moreover, even when connecting to a wrapper, the communication will still be done over the Ditto REST API, so the Ditto prefix does not seem entirely out of place ;)

e-grigorov replied:
I wasn't able to check the linked PR but generally speaking, it's about cloud2edge package setup. This PR is about generic integration setup which should be able to run with any backend setup (including cloud2edge package).

nzhechev-bosch replied:
Sure, with any backend setup, that supports Ditto.  My point is still valid.

dimiter-todorov-georgiev replied:
I have to agree with Nikola here, the client API we're using is explicitly called Ditto.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,430 @@
+
+const (
+	envVariablesPrefix    = "SCT"
+	featureName           = "ConnectorTestFeature"
```
This is a feature identifier, not a name.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,430 @@
+	case result := <-ch:
+		return result.cfg, result.err
+	case <-time.After(timeout):
+		return nil, fmt.Errorf("thing confing not received in %v", timeout)
```
typo: confing -> config

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,430 @@
+			Headers: headers,
+			Path:    strings.Replace(msg.Path, "/inbox/", "/outbox/", 1),
+			Value:   response,
+			Status:  http.StatusNoContent,
```
If response value is required, StatusOK should be used.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,430 @@
+	featureName           = "ConnectorTestFeature"
+	propertyName          = "testProperty"
+	commandName           = "testCommand"
+	commandResponseFormat = "response[%s]"
```
The response format can be Ditto compliant.

nzhechev-bosch replied:
According to the [API spec](https://www.eclipse.org/ditto/basic-messages.html#payload), Ditto does not care what the message payload is.

dimiter-todorov-georgiev replied:
@e-grigorov What do you suggest we should do here? We can leave as-is for now.

e-grigorov replied:
@dimiter-todorov-georgiev - as this approach doesn't break Ditto spec, no updates are required. I'll resolve the conversation.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,430 @@
+		}
+
+		if err := dittoClient.Reply(requestID, reply); err != nil {
+			suite.T().Error("failed to send response:", err)
```
The assertions can be unified. For example,  `require.NoError(suite.T(), err, "initialize ditto client")` is used.

nzhechev-bosch replied:
This call is in goroutine. Calling require.NoError will probably have unexpected consequences, as it invokes TestingT.failNow().

dimiter-todorov-georgiev replied:
Not being able to send a response will ultimately end up in failing the Command test because it expects a response within a certain timeout. So a failure here will also fail the actual test as well, with the code staying as-is. I think we can keep it this way.

e-grigorov replied:
@dimiter-todorov-georgiev - I cannot understand your comment here. Could you please clarify how it's related with the topic. I completely understand the @nzhechev-bosch concerns and I agree but I cannot understand your points.

dimiter-todorov-georgiev replied:
This is an async callback which will be called outside of our test case methods. If this callback fails in anything it's trying to do, then the actual test expecting the feature to send a response back to the backend will fail (the testCommand test method)

e-grigorov replied:
Yes, it'll fail but we cannot bypass the test message because the user will not be able to understand the reason behind this fail.

dimiter-todorov-georgiev replied:
I think we can fail the tests asynchronously, I will research to see how we should do it properly.

e-grigorov replied:
Could you please share the research? We ended with FailNow in the current source code state where we have API doc statement:
```
FailNow must be called from the goroutine running the test or benchmark function, not from other goroutines created during the test.
```

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,430 @@
+	suite.T().Logf("current time: %v", time.Now())
+
+	delta := int64(suite.cfg.TimeDeltaMs)
+	assert.Less(suite.T(), status.ReadySince.UnixMilli(), time.Now().UnixMilli()+delta, "readySince should be BEFORE current time")
```
What's the value to have a delta here?

nzhechev-bosch replied:
In some cases when the suite connector is restarted - e.g. to reload its configuration - and the test is started right away, it is possible for readySince to be reported as few milliseconds after time.Now(). The delta prevents the test from failing in such occurrences.

dimiter-todorov-georgiev replied:
I think we should keep this code because the connection status only contains "ready since" and "ready until". This algorithm is a good sanity check for "ready since".

e-grigorov replied:
@dimiter-todorov-georgiev - initially, there was no loop to check that the device is online and so called `sanity check` was an issue. The reason was very simple, it's possible to have a running TC but offline device i.e. most probably all assertions will fail. This was fixed with the outer loop and I guess that the delta is not needed any more or I am missing something?

dimiter-todorov-georgiev replied:
I think it's more of a way to have some wiggle room for the client and server having slightly different clocks. In my mind, this is like comparing floats, you always want to have some epsilon instead of comparing directly, due to inaccuracies.

e-grigorov replied:
If that the case, I could agree but some details have to be clarified. Why are 5 sec used as a default value? I guess that by default the delta should be 0 sec and if there is clock diff, it can be specified. Does it make sense?

dimiter-todorov-georgiev replied:
I think I understand. We can make this value configurable e.g. through an environment variable.

e-grigorov replied:
What's the status here? Now, there is an environment variable but the default value is still 5 sec. Are we going to set it to 0 sec?

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,430 @@
+
+	delta := int64(suite.cfg.TimeDeltaMs)
+	assert.Less(suite.T(), status.ReadySince.UnixMilli(), time.Now().UnixMilli()+delta, "readySince should be BEFORE current time")
+	assert.Greater(suite.T(), status.ReadySince.UnixMilli(), time.Now().UnixMilli()-delta, "readyUntil should be AFTER current time")
```
readyUntil should be used here and it can be asserted with 9999-12-31T23:59:59Z because the device has to be online.

nzhechev-bosch committed:
[#44] Replaced Ditto host/port config properties with single address property

Signed-off-by: Nikola Zhechev <Nikola.Zhechev@bosch.io>

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,430 @@
+	namespace := model.NewNamespacedIDFrom(suite.thingCfg.DeviceID)
+	cmd.Topic = fmt.Sprintf("%s/%s/things/live/messages/%s", namespace.Namespace, namespace.Name, commandName)
+	cmd.Path = fmt.Sprintf("/features/%s/inbox/messages/%s", featureName, commandName)
+	cmd.Value = "request"
```
The request can be Ditto compliant.

nzhechev-bosch replied:
According to the [API spec](https://www.eclipse.org/ditto/basic-messages.html#payload), Ditto does not care what the message payload is.

dimiter-todorov-georgiev replied:
@e-grigorov What do you suggest we should do here? We can leave as-is for now.

e-grigorov replied:
@dimiter-todorov-georgiev - as this approach doesn't break Ditto spec, no updates are required. I'll resolve the conversation.

nzhechev-bosch committed:
[#44] Minor adjustments, based on PR comments:
New Eclipse Kanto license header used
Renamed featureName to featureID
StatusOK used for command response
readyUntil now expected to be "9999-12-31T23:59:59Z"

Signed-off-by: Nikola Zhechev <Nikola.Zhechev@bosch.io>

nzhechev-bosch committed:
[#44] Now the test waits some time for the connection status to be updated

Signed-off-by: Nikola Zhechev <Nikola.Zhechev@bosch.io>

nzhechev-bosch committed:
[#44] Status timeout name fixed

Signed-off-by: Nikola Zhechev <Nikola.Zhechev@bosch.io>

dimiter-todorov-georgiev commented on integration/connector_test.go:
```diff
@@ -0,0 +1,470 @@
+	suite.T().Log(getConfigHelp(*cfg, envVariablesPrefix))
+
+	if err := initConfigFromEnv(cfg, envVariablesPrefix); err != nil {
+		suite.T().Skip(err)
```
If run with `go test` in this directory, this gets an error:
`Error initializing from config: env variable SCT_DITTO_ADDRESS not set`
and skips the test, declaring that it's successful.
I think we should fail the test instead.
`assert.NoErr(t, err).Fatal()`

nzhechev-bosch replied:
I beg to disagree. The test could be run as part of a set of tests - e.g. all suite connector tests - in an environment where cloud infrastructure is not available. It makes no sense to fail in that scenario.

More generally, this is indication of missing test precondition, not a bug in the tested functionality.

dimiter-todorov-georgiev replied:
If we automate this test in a properly setup environment, how do we distinguish between "preconditions failed" vs "test run successfully"? 

If the test fails on preconditions setup, then we can simply skip running it in a non-proper environment.

As it is, it can silently get skipped in an actual test, in case we got the environment wrong.

If the test is run manually, then it might be helpful to log the error, e.g. 
`fmt.Printf("Skipping test because of error: %v\n", err)`

nzhechev-bosch replied:
> If we automate this test in a properly setup environment, how do we distinguish between "preconditions failed" vs "test run successfully"?

In the first case - the particular test will have `SKIP` status in the final report, in the second case it will have `PASS` status.

>As it is, it can silently get skipped in an actual test, in case we got the environment wrong.

It is not skipped silently - as you mentioned, a message denoting exactly which environmental variable is missing is displayed.

If your objection is, that the message can be misleading, because it contains the word "error", I can agree to that. Maybe a more suitable message would be `Could not initialize config - env variable SCT_DITTO_ADDRESS  not set`

dimiter-todorov-georgiev replied:
I still think that this should be a fail because right now, on a specific missing env variable, the test isn't run at all but the reported status is success (return code of 0). Otherwise, invoking bash scripts should check the output for the SKIP word, which is brittle.

refi64 replied:
Could the new tests just use build tags? i.e. `go test` would *skip* these integration tests by default, but `go test --tags=integration` would only run the integration tests, failing hard if the environment setup isn't available.

dimiter-todorov-georgiev replied:
Done. We used the tag e2einttest, because integration is already being used for something else in the Kanto codebase.

dimiter-todorov-georgiev replied:
We have decided to use the tag `integration`.

refi64 commented on integration/connector_test.go:
```diff
@@ -0,0 +1,470 @@
+}
+
+func (suite *ConnectorSuite) TestCommand() {
+	ws, err := suite.newWSConnection()
```
Should this & the newWSConnection call below check the error value to print it?

nzhechev-bosch replied:
Yes, absolutely.

nzhechev-bosch committed:
[#44] Requre no error on websocket connect

Signed-off-by: Nikola Zhechev <Nikola.Zhechev@bosch.io>

refi64 commented on integration/connector_test.go:
```diff
@@ -0,0 +1,472 @@
+	}
+
+	if url.Scheme == "https" {
+		return fmt.Sprintf("wss://%s:%s", url.Hostname(), url.Port()), nil
```
So I accidentally found out that `Port()` always returns `"` if a port isn't present. This is...slightly unfortunate for http & https URLs, I had set `SCT_DITTO_ADDRESS=https://ditto.apertis.dev/`, so this URL becomes `wss://ditto.apertis.dev:/`, which doesn't actually work. I think this line should have the port default to `443` for standard HTTPS, and the next use should be `80` for standard HTTP?

dimiter-todorov-georgiev replied:
Fixed, please take another look

Commit:
[#44] Require no error on setting up the test

Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Do not use a prefix for env variables

Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Address PR feedback - use default HTTPS and HTTP ports if missing

Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Use build tag e2einttest for the new integration tests. This allows the regular go test command to succeed by skipping the integration tests, while still allowing the tests to be built to a binary by using the command go test --tags=e2einttest . -c -o sc-test

    Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] The build tag e2einttest is now replaced by integration

Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,488 @@
+	"github.com/eclipse/ditto-clients-golang/model"
+	"github.com/eclipse/ditto-clients-golang/protocol"
+	"github.com/eclipse/ditto-clients-golang/protocol/things"
+	mqtt "github.com/eclipse/paho.mqtt.golang"
```
mqtt should be MQTT - an example can be found [here](https://github.com/eclipse/paho.mqtt.golang/blob/master/cmd/sample/main.go). It goes as [an initialism](https://github.com/golang/go/wiki/CodeReviewComments#initialisms). It's a general issue and can be fixed in all code snippets.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,488 @@
+	PolicyID string `json:"policyId"`
+}
+
+type commandPayload struct {
```
Can we switch to Ditto Go library instead of these structures and hardcoded strings?

dimiter-todorov-georgiev replied:
I will research this topic.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,488 @@
+	Value   string                 `json:"value"`
+}
+
+type commandResponsePayload struct {
```
Can we switch to Ditto Go library instead of these structures and hardcoded strings?

dimiter-todorov-georgiev replied:
I will research this topic.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,488 @@
+func getPortOrDefault(url *url.URL) string {
+	port := url.Port()
+	if port == "" {
+		if url.Scheme == https {
```
We have this check in the invoker. Can we find a more elegant solution?

dimiter-todorov-georgiev replied:
This function is intended to be stand-alone, to be callable from any context.

e-grigorov replied:
This is a private function which is used only by the integration TC. FMPOV, it's not callable from any context.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,488 @@
+	suite.T().Log(getConfigHelp(*cfg))
+
+	if err := initConfigFromEnv(cfg); err != nil {
+		suite.T().Fail()
```
What is the reason to have Fail and FailNow right after?

dimiter-todorov-georgiev replied:
Fail marks the current test as failed, and Fatal exits the test run, preventing any more tests from running. I suppose we could only use Fatal here.

dimiter-todorov-georgiev replied:
I am going to remove .Fail() and only keep Fatalf() (which calls FailNow()).

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,488 @@
+
+	suite.T().Logf("thing config: %+v", *thingCfg)
+
+	dittoClient, err := ditto.NewClientMQTT(mqttClient, ditto.NewConfiguration())
```
Can we omit the configuration, just because the default one is used i.e. there are no customizations?

dimiter-todorov-georgiev replied:
I will research this topic.

dimiter-todorov-georgiev replied:
ditto.NewConfiguration() sets some default values to the configuration. But on the other hand, the ditto.NewClientMQTT function allows a nil configuration. I am not sure which is better, we can probably keep it as it is. so we can get the default values.

e-grigorov replied:
If nil is set to ditto.NewClientMQTT, is the default used?

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,488 @@
+
+	dittoClient.Subscribe(func(requestID string, msg *protocol.Envelope) {
+		if msg.Path != fmt.Sprintf("/features/%s/inbox/messages/%s", featureID, commandName) {
+			suite.T().Logf("unexpected command: %s", msg.Path)
```
On unexpected commands, do we need just log or TC should fail?

dimiter-todorov-georgiev replied:
This is an async callback which will be called outside of our test case methods. If this callback fails in anything it's trying to do, then the actual test expecting the feature to send a response back to the backend will fail (the testCommand test method)

dimiter-todorov-georgiev replied:
I think it will be OK to use suite.T().Fatalf(...) here. This will also fail the whole test run.

e-grigorov replied:
> I think it will be OK to use suite.T().Fatalf(...) here. This will also fail the whole test run.

Does it work in this asynch case?

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,488 @@
+
+		response := fmt.Sprintf(commandResponseFormat, value)
+
+		reply := &protocol.Envelope{
```
For this reply, [the builder can be used](https://github.com/eclipse-ditto/ditto-clients-golang#subscribing-and-handling-messages).

dimiter-todorov-georgiev replied:
I will research this topic.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,488 @@
+		}
+
+		delta := int64(suite.cfg.TimeDeltaMs)
+		assert.Less(suite.T(), status.ReadySince.UnixMilli(), time.Now().UnixMilli()+delta, "readySince should be BEFORE current time")
```
`BEFORE` - I guess that there is no real need for uppercase here.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,488 @@
+
+	err = suite.sendDittoEvent(channel, msg)
+
+	require.NoError(suite.T(), err)
```
General issue, it's always a better option to use the version with a message.

dimiter-todorov-georgiev replied:
Do you mean MQTT messages?

e-grigorov replied:
No, I mean that it's more useful to use: `require.NoError(suite.T(), err, "something useful here")` instead of `require.NoError(suite.T(), err)`

Commit:
[#44] Changes to environment variables according to feedback
- Rename Hono environment variables to DeviceRegistryApi
- Rename Ditto environment variables to DigitalTwinApi
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Code changes due to PR feedback
- Renamed mqtt package import to MQTT
- Removed unneeded call to .Fail()
- Properly failing the test when responses are unable to be sent to the backend
- Refactored getPortOrDefault()
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Integration test bugix
- Fix error handling when receiving an unknown command. This can happen for reasons which are unrelated to this test, so instead of failign the test, log the unrecognized command and continue the test.
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Changes due to PR comments
- Use the Ditto library where appropriate, removing custom structs
- Use constructors where required
- Use proper error messages with require.NoError() where they were missing
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Fix golint issues
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Remove an accidentally added file
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Remove asynchronous test failures, move them to the test goroutine
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,506 @@
+	defer suite.dittoClient.Unsubscribe(dittoHandler)
+
+	namespace := model.NewNamespacedIDFrom(suite.thingCfg.DeviceID)
+	cmd := things.NewCommand(namespace)
```
It's not a command but message has to be used. Switching to Ditto Go library, the code should become: `things.NewMessage(namespace).Feature(featureID).Inbox(commandName)...`

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,506 @@
+	timeout := time.Duration(suite.cfg.StatusTimeoutMs * int(time.Millisecond))
+	threshold := time.Now().Add(timeout)
+
+	for {
```
Usually, delay is used between loop iterations.

Commit:
[#44] Address code review feedback
- Use 0 as default offset when comparing connection status' readySince to now
- Added a sleep time between attempts to read connection status, defaulting to 2 seconds
- Changed command test code to properly use the Message struct from the ditto library
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Add a missing return statement
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Fix a golint issue: remove the new lines at the end of errors
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Refactor beginWSWait in preparation for the common integration test code
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Fix issues in ProcessMessages and rename to ProcessWSMessages
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Remove the Subscribe and WaitSubscribeResult functions
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Unsubscribe from local MQTT events on getThingConfig()
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

refi64 commented on integration/connector_test.go:
```diff
@@ -0,0 +1,522 @@
+		ReadyUntil time.Time `json:"readyUntil"`
+	}
+
+	timeout := time.Duration(suite.cfg.StatusTimeoutMs * int(time.Millisecond))
```
As it turns out, these are prone to overflow on 32-bit systems: `time.Millisecond` is in nanoseconds, so some of these can very easily pass the limits of an int. (e.g. the MQTT event timeout is 30 billion nanoseconds, and the status timeout is 10 billion). This results in some very nice errors:

```
2022-11-09 20:45:20.517717   OUT =>             Error:          Received unexpected error:
2022-11-09 20:45:20.518031   OUT =>                             WS response not received in -64.771072ms, last error: <nil>
```

Patch to fix:

```
diff --git a/integration/connector_test.go b/integration/connector_test.go
index 7754d0e..21e15a6 100644
--- a/integration/connector_test.go
+++ b/integration/connector_test.go
@@ -164,11 +164,11 @@ func (suite *ConnectorSuite) TestConnectionStatus() {
 		ReadyUntil time.Time `json:"readyUntil"`
 	}
 
-	timeout := time.Duration(suite.cfg.StatusTimeoutMs * int(time.Millisecond))
+	timeout := time.Duration(suite.cfg.StatusTimeoutMs) * time.Millisecond
 	threshold := time.Now().Add(timeout)
 
 	firstTime := true
-	sleepDuration := time.Duration(suite.cfg.TimeSleepMs * int(time.Millisecond))
+	sleepDuration := time.Duration(suite.cfg.TimeSleepMs) * time.Millisecond
 	for {
 		if !firstTime {
 			time.Sleep(sleepDuration)
@@ -283,7 +283,7 @@ func (suite *ConnectorSuite) TestCommand() {
 	err = websocket.JSON.Send(ws, cmdMsgEnvelope)
 	require.NoError(suite.T(), err, "unable to send command to the backend via websocket")
 
-	commandTimeout := time.Duration(suite.cfg.EventTimeoutMs * int(time.Millisecond))
+	commandTimeout := time.Duration(suite.cfg.EventTimeoutMs) * time.Millisecond
 	timeoutCh := beginWait(commandTimeout, commandResponseCh, func() {})
 	require.NoError(suite.T(), <-timeoutCh, "command should be received and response should be sent")
 	require.NoError(suite.T(), <-respCh, "command response should be received")
@@ -360,7 +360,7 @@ func (suite *ConnectorSuite) sendDittoEvent(topic string, message interface{}) e
 		return err
 	}
 	token := suite.mqttClient.Publish(topic, 1, false, payload)
-	timeout := time.Duration(suite.cfg.MqttAcknowledgeTimeoutMs * int(time.Millisecond))
+	timeout := time.Duration(suite.cfg.MqttAcknowledgeTimeoutMs) * time.Millisecond
 	if !token.WaitTimeout(timeout) {
 		return ditto.ErrAcknowledgeTimeout
 	}
@@ -384,7 +384,7 @@ func beginWait(timeout time.Duration, resultCh chan error, closer func()) chan e
 }
 
 func (suite *ConnectorSuite) beginWSWait(ws *websocket.Conn, check func(payload []byte) error) chan error {
-	timeout := time.Duration(suite.cfg.EventTimeoutMs * int(time.Millisecond))
+	timeout := time.Duration(suite.cfg.EventTimeoutMs) * time.Millisecond
 	resultCh := make(chan error)
 
 	go func() {

```

Commit:
[#44] Consistently use MillisToDuration() and switch to 64-bit math to avoid overflow
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Use the common integration test repository
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Remove redundant log messages
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

hristobojilov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,317 @@
+
+type ConnectorSuite struct {
+	suite.Suite
+	suiteInitializer *util.SuiteInitializer
```
Use golang struct embedding and don't use pointers here. All members of SuiteInitializer are pointers deliberately to make it possible to use this without external pointers.

hristobojilov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,317 @@
+	assert.Equal(suite.T(), fmt.Sprintf("\"%s\"", newValue), strings.TrimSpace(string(body)), "property value updated")
+}
+
+func (suite *ConnectorSuite) sendDittoEvent(topic string, message interface{}) error {
```
Use SendMQTTMessage instead.

hristobojilov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,317 @@
+
+func (suite *ConnectorSuite) TestCommand() {
+	init := suite.suiteInitializer
+	cfg := init.Cfg
```
Use embedding

Commit:
[#44] Use embedding for SuiteInitializer
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Use a common code function call
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,293 @@
+	defer suite.DittoClient.Unsubscribe(dittoHandler)
+
+	correlationID := uuid.New().String()
+	namespace := model.NewNamespacedIDFrom(suite.thingCfg.DeviceID)
```
This is not a namespace. That's the device identifier of type NamespacedID.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,293 @@
+	"golang.org/x/net/websocket"
+)
+
+type suiteConnectorTestConfig struct {
```
There are:
- util.TestConfiguration
- util.ThingConfiguration
- ConnectorSuite

i.e. Config should be replaced with Configuration and suite prefix should be removed just because the overlapping with the testify suite. Proposal: connectorTestConfiguration.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,293 @@
+	StatusTimeoutMs int `env:"SCT_STATUS_TIMEOUT_MS" envDefault:"10000"`
+
+	TimeDeltaMs int `env:"SCT_TIME_DELTA_MS" envDefault:"0"`
+	TimeSleepMs int `env:"SCT_TIME_SLEEP_MS" envDefault:"2000"`
```
TIME_SLEEP - there is no semantic in this name. Just an example, in the software update, there is similar property and it's called: downloadRetryInterval i.e. the interval between retries in case of failed download.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,293 @@
+type suiteConnectorTestConfig struct {
+	StatusTimeoutMs int `env:"SCT_STATUS_TIMEOUT_MS" envDefault:"10000"`
+
+	TimeDeltaMs int `env:"SCT_TIME_DELTA_MS" envDefault:"0"`
```
Similar issue as it's for TIME_SLEEP.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,293 @@
+	util.SuiteInitializer
+
+	thingCfg   *util.ThingConfiguration
+	testConfig *suiteConnectorTestConfig
```
We have thingCfg and testConfig. Could you please unify the names? For example, use cfg everywhere.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,293 @@
+func (suite *ConnectorSuite) SetupSuite() {
+	suite.Setup(suite.T())
+
+	cfg := suite.Cfg
```
Is there a real need to have this variable?

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,293 @@
+
+	testConfig := &suiteConnectorTestConfig{}
+	opts := env.Options{RequiredIfNoDef: true}
+	require.NoError(suite.T(), env.Parse(testConfig, opts), "Failed to process environment variables")
```
- Failed -> failed
- `process environment variables` - here, we have the context, we can add that these are suite connector env vars

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,293 @@
+		ReadyUntil time.Time `json:"readyUntil"`
+	}
+
+	cfg := suite.Cfg
```
Do we need a variable, just to pass it to a single method? It seems like a general topic. Could you please revise the usage?

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,293 @@
+		require.NoError(suite.T(), err, "connection status should be parsed")
+
+		suite.T().Logf("%+v", status)
+		suite.T().Logf("current time: %v", time.Now())
```
If we would like to make this log usable, it has to be assigned to a variable and after that this variable has to be used in <newVar>.Before(threshold) comparison. The issue is that two execution to Now() method will return two different values.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,293 @@
+			responsePayload := fmt.Sprintf(responsePayloadTemplate, value)
+			response := things.NewMessage(model.NewNamespacedID(msg.Topic.Namespace, msg.Topic.EntityName))
+			// respond to the message by using the outbox
+			path := strings.Replace(msg.Path, "/inbox/", "/outbox/", 1)
```
We don't need this raw string operation. We should use Feature(...) and Outbox(...) functions of the message.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,293 @@
+				protocol.WithContentType("application/json")).
+				WithTopic(msg.Topic).
+				WithPath(path).
+				WithValue(responsePayload).
```
Use the Message API instead.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,293 @@
+		return nil
+	}
+
+	err = websocket.JSON.Send(ws, cmdMsgEnvelope)
```
What's the reason to use WebSocket for commands? There is [REST API](https://www.eclipse.org/ditto/http-api-doc.html#/Messages).

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,293 @@
+	timeout := util.MillisToDuration(cfg.WsEventTimeoutMs)
+
+	// Check the sending of the response from the feature to the backend
+	var responseSentResult error
```
Why we have this check? The response check on the backend should be enough.

e-grigorov replied:
I think that this comment is still valid.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,293 @@
+		if err := checkEqual(respMsg.Status, http.StatusOK, "http ok"); err != nil {
+			return false, err
+		}
+		if err := checkEqual(fmt.Sprintf(responsePayloadTemplate, cmdMsgEnvelope.Value.(string)), respMsg.Value, "response payload"); err != nil {
```
responsePayload can be reused here.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,293 @@
+
+	require.NoError(suite.T(), util.WaitForWSMessage(cfg, ws, subAck), "acknowledgement %v should be received", subAck)
+
+	namespace := model.NewNamespacedIDFrom(suite.thingCfg.DeviceID)
```
This is not a namespace.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,293 @@
+	TimeSleepMs int `env:"SCT_TIME_SLEEP_MS" envDefault:"2000"`
+}
+
+type ConnectorSuite struct {
```
Do we need this as a public struct?

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,293 @@
+
+	suite.thingCfg = thingCfg
+	suite.testConfig = testConfig
+	suite.thingURL = fmt.Sprintf("%s/api/2/things/%s", strings.TrimSuffix(cfg.DigitalTwinAPIAddress, "/"), thingCfg.DeviceID)
```
strings.TrimSuffix - is it a common logic, can we move it to the integration util?

Commit:
[#44] Address code review feedback
- Switch to using the Digital Twin REST API for sending a command to the feature
- Minor fixes
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Code cleanup
- Removed unneeded lines
- Use the DITTO API for checking of the expected topic
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Refactored code to use the new common integration test logic
- Removed variables which come from the common code
- Use the new API functions provided by the common logic
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Removed debug prints
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

hristobojilov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,265 @@
+
+	featurePropertyPath := util.GetFeaturePropertyPath(featureID, propertyName)
+	result := util.ProcessWSMessages(suite.Cfg, ws, func(msg *protocol.Envelope) (bool, error) {
+		expectedTopic := protocol.Topic{
```
Use the new utility method GetTwinEventTopic if possible

Commit:
[#44] Use util.GetTwinEventTopic
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,258 @@
+		}
+		firstTime = false
+
+		statusURL := fmt.Sprintf("%s/features/ConnectionStatus/properties/status", suite.thingURL)
```
It's a simple string concatenation, there is no need to have more complex formatting.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,258 @@
+	}
+
+	dittoHandler := func(requestID string, msg *protocol.Envelope) {
+		expectedPath := strings.ReplaceAll(util.GetFeatureOutboxMessagePath(featureID, commandName), "outbox", "inbox")
```
That's not the way how this should work.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,258 @@
+func (suite *connectorSuite) TestCommand() {
+	commandResponseCh := make(chan error)
+
+	responsePayload := func(value string) string {
```
Function of 1 line of code that is assigned to a variable for a simple printf. This is very untypical usage.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,258 @@
+		require.NoError(suite.T(), err, "connection status should be parsed")
+
+		suite.T().Logf("%+v", status)
+		suite.T().Logf("current time: %v", now)
```
`now` is initialized only in case of an error, that's an issue.

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,258 @@
+		suite.T().Logf("current time: %v", now)
+
+		forever := time.Date(9999, time.December, 31, 23, 59, 59, 0, time.UTC)
+		if status.ReadyUntil != forever {
```
It's most appropriate to use Equal instead as it's stated in the API doc: https://pkg.go.dev/time#Time

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,258 @@
+	const commandPayload = "request"
+
+	commandTimeoutSeconds := suite.connectorTestCfg.CommandTimeoutMs / 1000
+	digitalTwinAPI := fmt.Sprintf("%s/inbox/messages/%s?timeout=%d", suite.featureURL, commandName, commandTimeoutSeconds)
```
I would suggest to switch to the new utility test function: ExecuteOperation

e-grigorov commented on integration/connector_test.go:
```diff
@@ -0,0 +1,258 @@
+
+	require.NoError(suite.T(), err, "unable to send event to the backend")
+
+	featurePropertyPath := util.GetFeaturePropertyPath(featureID, propertyName)
```
The name has to be aligned with expectedTopic variable and the definition has to be moved to the callback, on the sam elevel as expectedTopic.

e-grigorov commented on go.mod:
```diff
@@ -22,17 +24,17 @@ require (
 require (
 	github.com/davecgh/go-spew v1.1.1 // indirect
-	github.com/google/uuid v1.1.1 // indirect
+	github.com/google/uuid v1.3.0 // indirect
```
It's a runtime dependency and the notice has to be updated also.

Commit:
[#44] Addressing PR feedback:
- Update NOTICE.md with the updated version of the uuid library.
- Remove the extra check (and relevant code) for sending an answer to an operation message.
- Use the util function ExecuteOperation.
- Compare time.Time using time.Equal.
- Fix the initialization of the variable now.
- Move the initialization of a variable closer to where it's used, reducing its scope.
- Stop using GetFeatureOutboxMessagePath to obtain the inbox.
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Use the new utility functions GetFeaturePropertyValue and GetFeatureInboxMessagePath.
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

Commit:
[#44] Remove an unused environment variable.
Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>
